File: src/mock/state.ts
/**
 * Mutable mock data backing the mock command layer (src/mock/core.ts).
 *
 * This stands in for the Rust backend's state. The UI treats it as read-only
 * data it would normally fetch via Tauri commands; only the simulated "model
 * download" flow mutates it (via markModelDownloaded / setCurrentModel).
 */
import type {
  AppSettings,
  AudioDevice,
  HistoryEntry,
  ModelInfo,
} from "@/bindings";

function makeSettings(): AppSettings {
  return {
    bindings: {
      transcribe: {
        id: "transcribe",
        name: "Transcribe",
        description: "Start recording and transcribe audio",
        default_binding: "cmd+shift+space",
        current_binding: "cmd+shift+space",
      },
      cancel: {
        id: "cancel",
        name: "Cancel",
        description: "Cancel the current recording",
        default_binding: "escape",
        current_binding: "escape",
      },
    },
    push_to_talk: false,
    audio_feedback: true,
    audio_feedback_volume: 50,
    sound_theme: "marimba",
    start_hidden: false,
    autostart_enabled: false,
    update_checks_enabled: true,
    selected_model: undefined,
    always_on_microphone: false,
    selected_microphone: "Default",
    clamshell_microphone: "Default",
    selected_output_device: "Default",
    translate_to_english: false,
    selected_language: "auto",
    overlay_position: "bottom",
    debug_mode: false,
    log_level: "info",
    custom_words: [],
    model_unload_timeout: "min_5",
    word_correction_threshold: 60,
    history_limit: 100,
    recording_retention_period: "weeks_2",
    paste_method: "ctrl_v",
    clipboard_handling: "dont_modify",
    auto_submit: false,
    auto_submit_key: "enter",
    post_process_enabled: false,
    post_process_provider_id: "openai",
    post_process_providers: [
      {
        id: "openai",
        label: "OpenAI",
        base_url: "https://api.openai.com/v1",
        models_endpoint: "/models",
        supports_structured_output: true,
      },
      {
        id: "groq",
        label: "Groq",
        base_url: "https://api.groq.com/openai/v1",
        models_endpoint: "/models",
      },
      {
        id: "custom",
        label: "Custom",
        base_url: "",
        allow_base_url_edit: true,
      },
    ],
    post_process_api_keys: {},
    post_process_models: {},
    post_process_prompts: [],
    post_process_selected_prompt_id: null,
    mute_while_recording: false,
    append_trailing_space: true,
    app_language: "en",
    experimental_enabled: false,
    lazy_stream_close: false,
    keyboard_implementation: "tauri",
    show_tray_icon: true,
    paste_delay_ms: 0,
    typing_tool: "auto",
    external_script_path: null,
    custom_filler_words: null,
    whisper_accelerator: "auto",
    ort_accelerator: "auto",
    whisper_gpu_device: 0,
    extra_recording_buffer_ms: 300,
  };
}

const models: ModelInfo[] = [
  {
    id: "small",
    name: "Whisper Small",
    description: "Fast and fairly accurate.",
    filename: "small.bin",
    url: "https://example.com/small.bin",
    sha256: null,
    size_mb: 488,
    is_downloaded: false,
    is_downloading: false,
    partial_size: 0,
    is_directory: false,
    engine_type: "Whisper",
    accuracy_score: 6,
    speed_score: 9,
    supports_translation: true,
    is_recommended: true,
    supported_languages: ["auto", "en", "ru"],
    supports_language_selection: true,
    is_custom: false,
  },
  {
    id: "turbo",
    name: "Whisper Turbo",
    description: "Balanced accuracy and speed.",
    filename: "turbo.bin",
    url: "https://example.com/turbo.bin",
    sha256: null,
    size_mb: 812,
    is_downloaded: false,
    is_downloading: false,
    partial_size: 0,
    is_directory: false,
    engine_type: "Whisper",
    accuracy_score: 8,
    speed_score: 7,
    supports_translation: true,
    is_recommended: false,
    supported_languages: ["auto", "en", "ru"],
    supports_language_selection: true,
    is_custom: false,
  },
  {
    id: "medium",
    name: "Whisper Medium",
    description: "Good accuracy, medium speed.",
    filename: "medium.bin",
    url: "https://example.com/medium.bin",
    sha256: null,
    size_mb: 1500,
    is_downloaded: false,
    is_downloading: false,
    partial_size: 0,
    is_directory: false,
    engine_type: "Whisper",
    accuracy_score: 8,
    speed_score: 5,
    supports_translation: true,
    is_recommended: false,
    supported_languages: ["auto", "en", "ru"],
    supports_language_selection: true,
    is_custom: false,
  },
];

const microphones: AudioDevice[] = [
  { index: "default", name: "Default", is_default: true },
  { index: "0", name: "MacBook Pro Microphone", is_default: false },
  { index: "1", name: "External Microphone", is_default: false },
];

const outputDevices: AudioDevice[] = [
  { index: "default", name: "Default", is_default: true },
  { index: "1", name: "Built-in Output", is_default: false },
];

const history: HistoryEntry[] = [
  {
    id: 1,
    file_name: "recording-1.wav",
    timestamp: 1718500000000,
    saved: true,
    title: "Meeting notes",
    transcription_text:
      "This is a mock transcription entry demonstrating the history view.",
    post_processed_text: null,
    post_process_prompt: null,
    post_process_requested: false,
  },
  {
    id: 2,
    file_name: "recording-2.wav",
    timestamp: 1718400000000,
    saved: false,
    title: "Voice memo",
    transcription_text:
      "Another mock entry so the history list is not empty in the UI demo.",
    post_processed_text: null,
    post_process_prompt: null,
    post_process_requested: false,
  },
];

export const state = {
  settings: makeSettings(),
  defaultSettings: makeSettings(),
  models,
  microphones,
  outputDevices,
  history,
  currentModel: "",
  downloaded: new Set<string>(),
};

export function markModelDownloaded(modelId: string): void {
  state.downloaded.add(modelId);
}

export function removeModel(modelId: string): void {
  state.downloaded.delete(modelId);
  if (state.currentModel === modelId) state.currentModel = "";
}

export function setCurrentModel(modelId: string): void {
  state.currentModel = modelId;
  state.settings.selected_model = modelId;
}

File: src-tauri/src/state.rs
//! In-process application state, mirroring `src/mock/state.ts`.
//!
//! This is the backing store for the logging stub. Getters read from here,
//! mutators no-op (or lightly mutate, as the mock does), and the simulated
//! model download flips entries in `downloaded`.

use std::collections::HashSet;

use crate::types::{
    AppSettings, AudioDevice, EngineType, HistoryEntry, ModelInfo, ShortcutBinding,
};

/// Raw model catalogue without the runtime `is_downloaded` flag, which is
/// stamped in from `downloaded` at read time (matching `mock/core.ts`).
#[derive(Debug, Clone)]
struct ModelRecord {
    id: String,
    name: String,
    description: String,
    filename: String,
    url: Option<String>,
    sha256: Option<String>,
    size_mb: i64,
    engine_type: EngineType,
    accuracy_score: i64,
    speed_score: i64,
    supports_translation: bool,
    is_recommended: bool,
    supported_languages: Vec<String>,
    supports_language_selection: bool,
}

impl ModelRecord {
    fn to_info(&self, is_downloaded: bool) -> ModelInfo {
        ModelInfo {
            id: self.id.clone(),
            name: self.name.clone(),
            description: self.description.clone(),
            filename: self.filename.clone(),
            url: self.url.clone(),
            sha256: self.sha256.clone(),
            size_mb: self.size_mb,
            is_downloaded,
            is_downloading: false,
            partial_size: 0,
            is_directory: false,
            engine_type: self.engine_type,
            accuracy_score: self.accuracy_score,
            speed_score: self.speed_score,
            supports_translation: self.supports_translation,
            is_recommended: self.is_recommended,
            supported_languages: self.supported_languages.clone(),
            supports_language_selection: self.supports_language_selection,
            is_custom: false,
        }
    }
}

#[derive(Debug)]
pub struct AppState {
    pub settings: AppSettings,
    pub default_settings: AppSettings,
    models: Vec<ModelRecord>,
    pub microphones: Vec<AudioDevice>,
    pub output_devices: Vec<AudioDevice>,
    pub history: Vec<HistoryEntry>,
    pub current_model: String,
    pub downloaded: HashSet<String>,
}

impl Default for AppState {
    fn default() -> Self {
        Self {
            settings: AppSettings::mock(),
            default_settings: AppSettings::mock(),
            models: mock_models(),
            microphones: mock_microphones(),
            output_devices: mock_outputs(),
            history: mock_history(),
            current_model: String::new(),
            downloaded: HashSet::new(),
        }
    }
}

impl AppState {
    /// Models with `is_downloaded` stamped from the `downloaded` set.
    pub fn models(&self) -> Vec<ModelInfo> {
        self.models
            .iter()
            .map(|m| m.to_info(self.downloaded.contains(&m.id)))
            .collect()
    }

    /// Look up a single model's info, or `None` if unknown.
    pub fn model_info(&self, model_id: &str) -> Option<ModelInfo> {
        self.models
            .iter()
            .find(|m| m.id == model_id)
            .map(|m| m.to_info(self.downloaded.contains(&m.id)))
    }

    pub fn model_size_mb(&self, model_id: &str) -> i64 {
        self.models
            .iter()
            .find(|m| m.id == model_id)
            .map(|m| m.size_mb)
            .unwrap_or(100)
    }

    pub fn mark_downloaded(&mut self, model_id: &str) {
        self.downloaded.insert(model_id.to_string());
    }

    pub fn remove_model(&mut self, model_id: &str) {
        self.downloaded.remove(model_id);
        if self.current_model == model_id {
            self.current_model.clear();
        }
    }

    pub fn set_current_model(&mut self, model_id: &str) {
        self.current_model = model_id.to_string();
        self.settings.selected_model = Some(model_id.to_string());
    }

    /// Update a shortcut binding's current binding; returns the stored binding.
    pub fn set_binding(&mut self, id: &str, binding: &str) -> Option<ShortcutBinding> {
        if let Some(entry) = self.settings.bindings.get_mut(id) {
            entry.current_binding = binding.to_string();
            Some(entry.clone())
        } else {
            None
        }
    }

    pub fn reset_binding(&mut self, id: &str) -> Option<ShortcutBinding> {
        if let Some(entry) = self.settings.bindings.get_mut(id) {
            entry.current_binding = entry.default_binding.clone();
            Some(entry.clone())
        } else {
            None
        }
    }
}

fn mock_models() -> Vec<ModelRecord> {
    let langs = vec!["auto".to_string(), "en".to_string(), "ru".to_string()];
    vec![
        ModelRecord {
            id: "small".to_string(),
            name: "Whisper Small".to_string(),
            description: "Fast and fairly accurate.".to_string(),
            filename: "small.bin".to_string(),
            url: Some("https://example.com/small.bin".to_string()),
            sha256: None,
            size_mb: 488,
            engine_type: EngineType::Whisper,
            accuracy_score: 6,
            speed_score: 9,
            supports_translation: true,
            is_recommended: true,
            supported_languages: langs.clone(),
            supports_language_selection: true,
        },
        ModelRecord {
            id: "turbo".to_string(),
            name: "Whisper Turbo".to_string(),
            description: "Balanced accuracy and speed.".to_string(),
            filename: "turbo.bin".to_string(),
            url: Some("https://example.com/turbo.bin".to_string()),
            sha256: None,
            size_mb: 812,
            engine_type: EngineType::Whisper,
            accuracy_score: 8,
            speed_score: 7,
            supports_translation: true,
            is_recommended: false,
            supported_languages: langs.clone(),
            supports_language_selection: true,
        },
        ModelRecord {
            id: "medium".to_string(),
            name: "Whisper Medium".to_string(),
            description: "Good accuracy, medium speed.".to_string(),
            filename: "medium.bin".to_string(),
            url: Some("https://example.com/medium.bin".to_string()),
            sha256: None,
            size_mb: 1500,
            engine_type: EngineType::Whisper,
            accuracy_score: 8,
            speed_score: 5,
            supports_translation: true,
            is_recommended: false,
            supported_languages: langs,
            supports_language_selection: true,
        },
    ]
}

fn mock_microphones() -> Vec<AudioDevice> {
    vec![
        AudioDevice {
            index: "default".to_string(),
            name: "Default".to_string(),
            is_default: true,
        },
        AudioDevice {
            index: "0".to_string(),
            name: "MacBook Pro Microphone".to_string(),
            is_default: false,
        },
        AudioDevice {
            index: "1".to_string(),
            name: "External Microphone".to_string(),
            is_default: false,
        },
    ]
}

fn mock_outputs() -> Vec<AudioDevice> {
    vec![
        AudioDevice {
            index: "default".to_string(),
            name: "Default".to_string(),
            is_default: true,
        },
        AudioDevice {
            index: "1".to_string(),
            name: "Built-in Output".to_string(),
            is_default: false,
        },
    ]
}

fn mock_history() -> Vec<HistoryEntry> {
    vec![
        HistoryEntry {
            id: 1,
            file_name: "recording-1.wav".to_string(),
            timestamp: 1_718_500_000_000,
            saved: true,
            title: "Meeting notes".to_string(),
            transcription_text:
                "This is a mock transcription entry demonstrating the history view.".to_string(),
            post_processed_text: None,
            post_process_prompt: None,
            post_process_requested: false,
        },
        HistoryEntry {
            id: 2,
            file_name: "recording-2.wav".to_string(),
            timestamp: 1_718_400_000_000,
            saved: false,
            title: "Voice memo".to_string(),
            transcription_text:
                "Another mock entry so the history list is not empty in the UI demo.".to_string(),
            post_processed_text: None,
            post_process_prompt: None,
            post_process_requested: false,
        },
    ]
}
